### PR TITLE
gRPC channel config set limits

### LINKED ISSUE #12540: Add data rate limit
data rate limit  etc

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 1fc8dc9f-9cc8-bda7-2cda-e9461f7c3a5d
* Version Independent ID: 85b23cfb-29a8-37f4-16f7-fdf6b0dde773
* Content: [gRPC for ASP.NET Core configuration](https://docs.microsoft.com/en-us/aspnet/core/grpc/configuration?view=aspnetcore-3.0)
* Content Source: [aspnetcore/grpc/configuration.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/grpc/configuration.md)
* Product: **aspnet-core**
* GitHub Login: @JamesNK
* Microsoft Alias: **jamesnk**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #12540
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/grpc/configuration?view=aspnetcore-3.0&branch=pr-en-us-12539)

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@luizcarlosfaria 
To review the doc build 
- download [this DB file](https://www.dropbox.com/s/bxomestammbzqye/1.zip?dl=0)
- unzip
- view HTML file in FF (Fire Fox). It best to view the doc with the widest window that has no left or right TOC (Table  of contents)

**Comment by luizcarlosfaria** [comment]
@Rick-Anderson thank you, but we have small issue.

Description:
ChannelOptions.ReceiveMaxMessageSize must be a same value of ChannelOptions.MaxSendMessageLength.

Changes need:
FROM
```csharp
options.ReceiveMaxMessageSize = 10 * 1024 * 1024;  // 10 megabytes
```
TO
```csharp
options.ReceiveMaxMessageSize = 2 * 1024 * 1024;  // 2 megabytes
```

File: aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    public class Startup2
+    {
+        // This method gets called by the runtime. Use this method to add services to the container.
+        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
+#region snippet
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddGrpc().AddServiceOptions<MyService>(options =>
+            {
+                options.ReceiveMaxMessageSize = 10 * 1024 * 1024;  // 10 megabytes
```
Comment (luizcarlosfaria) [inline aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs]: To be consistent between client and server configs, need change:
FROM
```csharp
options.ReceiveMaxMessageSize = 10 * 1024 * 1024;  // 10 megabytes
```
TO
```csharp
options.ReceiveMaxMessageSize = 2 * 1024 * 1024;  // 2 megabytes
```

**Comment by Rick-Anderson** [comment]
@luizcarlosfaria  Thanks!
Can you check my last commit?

**Comment by luizcarlosfaria** [comment]
> @luizcarlosfaria Thanks!
> Can you check my last commit?

Ok, it's ok for me!

**Comment by luizcarlosfaria** [comment]
Thank you.

File: aspnetcore/grpc/configuration.md [hunk 2]
```diff
@@ -18,31 +18,50 @@ The following table describes options for configuring gRPC services:
...
-    options.ReceiveMaxMessageSize = 10 * 1024 * 1024; // 10 megabytes
-});
-```
+[!code-csharp[](~/grpc/configuration/sample/GrcpService/Startup2.cs?name=snippet)]
+
+## Configure Kestrel options
+
+Kestrel server has configuration options that affect the behavior of gRPC on ASP.NET Core.
+
+### Request body data rate limit
```
Comment (JamesNK) [inline aspnetcore/grpc/configuration.md]: Remove this section. gRPC framework now does it for you automatically.

File: aspnetcore/grpc/configuration.md [hunk 3]
```diff
@@ -18,31 +18,50 @@ The following table describes options for configuring gRPC services:
...
+| `MaxSendMessageLength` | `ReceiveMaxMessageSize` |
+| `MaxReceiveMessageLength` | `SendMaxMessageSize` |
+
+<!-- What are important User Agent options ? 
+And some client specific configurations like 
+ -->
+
+The following highlighted code sets the server options listed in the preceding table:
+
+[!code-csharp[](~/grpc/configuration/sample/GrcpService/Startup2.cs?name=snippet)]
```
Comment (JamesNK) [inline aspnetcore/grpc/configuration.md]: Setting the options in `AddServiceOptions<MyService>` will only change the limits for that service (`MyService`). Setting them in AddGrpc will change them for all services (unless overridden).

IMO change this to set these values in AddGrpc and add a link to AddServiceOption and explain about also changing them per service.

File: aspnetcore/grpc/configuration.md [hunk 4]
```diff
@@ -18,31 +18,50 @@ The following table describes options for configuring gRPC services:
...
+
+The following code sets the client options listed in the preceding table:
+
+[!code-csharp[](~/grpc/configuration/sample/Program.cs?name=snippet&highlight=3-6)]
```
Comment (JamesNK) [inline aspnetcore/grpc/configuration.md]: This is fine but hopefully next week we'll have a new managed client checked in. Note that it won't support message size limits in preview 6.

**Comment by Rick-Anderson** [comment]
@luizcarlosfaria  please review.

File: aspnetcore/grpc/configuration/sample/GrcpService/Program.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Hosting;
+
+namespace GrcpService
+{
+    #region snippet
```
Comment (JamesNK) [inline aspnetcore/grpc/configuration/sample/GrcpService/Program.cs]: You can delete this file

File: aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        // This method gets called by the runtime. Use this method to add services to the container.
+        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
+#region snippet
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddGrpc().AddServiceOptions<MyService>(options =>
+            {
+                options.ReceiveMaxMessageSize = 3 * 1024 * 1024;
+                options.SendMaxMessageSize    = 3 * 1024 * 1024;
```
Comment (luizcarlosfaria) [inline aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs]: For didactic purpose, doesn't use same values for ReceiveMaxMessageSize and SendMaxMessageSize.

Comment (Rick-Anderson) [reply aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs]: @luizcarlosfaria  
Por favor, recomende valores apropriados.
Como é isso @mairaw?

Comment (luizcarlosfaria) [reply aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs]: If change to 
```csharp
options.ReceiveMaxMessageSize = 2 * 1024 * 1024;  // 2 megabytes
options.SendMaxMessageSize = 5 * 1024 * 1024;     // 5 megabytes 
```
we address this point. 

Thank you @Rick-Anderson

Comment (Rick-Anderson) [reply aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs]: @luizcarlosfaria você pode verificar se eu consertei todos os problemas `@ JamesNK` identificados?

@luizcarlosfaria can you check that I've fixed all the problems `@JamesNK` identified?

Comment (luizcarlosfaria) [reply aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs]: Hi @Rick-Anderson, sorry about my terrible english. 
You can write in english, it's not a problem for me, i can understand very well. Write and speak is always a problem, sorry about that.

I'm working on your request now.

**Comment by Rick-Anderson** [comment]
@luizcarlosfaria  your english is fine. I just like to use the google translator.

**Comment by luizcarlosfaria** [comment]
Hi,
i'm looking the final version of wiki and the result doesn't have a topic about Client Configuration.

We have one topic "Configure services options",  but we don't have any topic like "Configure a client options" to explain with a table the options on client, it's ok?

**Comment by Rick-Anderson** [comment]
@luizcarlosfaria  added `## Configure client options`

**Comment by luizcarlosfaria** [comment]
Thank you @Rick-Anderson! How can i see the final result?

**Comment by Rick-Anderson** [comment]
https://www.dropbox.com/sh/yy994551cje0kap/AAB0tRnm7gze2kYrSYPSFHIda?dl=0

**Comment by luizcarlosfaria** [comment]
Thank you, thanks about your time.
 It's ok for me! All things are addressed.

**Comment by Rick-Anderson** [comment]
@JamesNK  you OK with publishing?

File: aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs (hunk 6 above)
Comment (mairaw) [reply aspnetcore/grpc/configuration/sample/GrcpService/Startup2.cs]: Ah sorry I was late to help with translations (was out on vacation) but I'm glad you guys figured out and got it done! 👏

**Pull request merged**